gnovm/pkg/gnolang/garbage_collector.go

```diff
@@ -42,65 +74,108 @@ func (m *Machine) GarbageCollect() (left int64, ok bool) {
 	}
 
 	// Visit exceptions
-	for exception := range m.Exceptions {
-		// XXX implement for exceptions.
-		// XXX Exception is not an object,
-		// so implement a custom method and pass in vis.
+	for _, exception := range m.Exceptions {
+		stop = exception.Visit(vis, m.Store)
+		if stop {
+			return -1, false
+		}
 	}
 
 	// Return bytes remaining.
 	maxBytes, bytes := m.Alloc.Status()
 	return maxBytes - bytes, true
 }
 
-// Returns a visitor that bumps the gcCycle counter
+// Returns a visitor that bumps the GCCycle counter
 // and stops if alloc is out of memory.
-func GCVisitorFn(gcCycle int64, alloc *Alloc) Visitor {
-	vis := func(o Object) bool {
-		// Return if already measured.
-		if o.GetLastGCCycle() == gcCycle {
-			return false // but don't stop
+func GCVisitorFn(gcCycle int64, alloc *Allocator) Visitor {
+	var vis func(value Value) bool
+
+	vis = func(v Value) bool {
+		if v == nil || v == (*Block)(nil) || v == (*PackageValue)(nil) {
```

Comment:
Can you rebase to the heap PR and see if this is still necessary? 
We should not have to check for typed nils.

Reply:
fixed.